backend/lib/index.js

```diff
@@ -181,15 +181,24 @@ module.exports = class GitSheets {
     return Promise.all(pendingDiffs)
   }
 
-  async merge (srcRef, dstRef) {
+  async merge (srcRef, dstRef, msg = null) {
     try {
       await this.git.mergeBase({'is-ancestor': true}, srcRef, dstRef)
     } catch (err) {
       throw new Error(`${srcRef} is not an ancestor of ${dstRef}`)
     }
+    const commitMsg = msg || `Merge ${dstRef}`
+    const dstTree = await this.repo.createTreeFromRef(dstRef)
+    const dstTreeHash = await dstTree.getHash()
+    const srcCommitHash = await this.git.revParse({verify: true}, srcRef)
+    const dstCommitHash = await this.git.revParse({verify: true}, dstRef)
+    const mergeCommitHash  = await this.git.commitTree(dstTreeHash, {
+      p: [srcCommitHash, dstCommitHash],
+      m: commitMsg
+    })
+
     const qualifiedSrcRef = await this.getQualifiedRef(srcRef)
-    const qualifiedDstRef = await this.getQualifiedRef(dstRef)
-    await this.git.updateRef(qualifiedSrcRef, qualifiedDstRef)
+    await this.git.updateRef(qualifiedSrcRef, mergeCommitHash)
```

Comment:
@timwis you can just pass `srcCommitHash` as a 3rd param to `update-ref`, this is the intended mechanism for making sure a target ref hasn't evolved while you've been building an addition to it. This is considered to be the "safe" way to write to a ref, and git's internal mechanics ensures that it is safe/consistent

Your current method would allow history to be destroyed (i.e. any history appended to the same `srcRef`), as I believe the guard will allow it to overwrite any ancestor, and doesn't protect from overwriting external changes that happen between the start and the end. 

I think we want both, as your first test ensures that we're fast-forwarding and don't need to worry about merges yet, but then passing the `oldValue` to `update-ref` ensures that is still the case by the time the new ref value actually gets applied to disk

Reply:
Oh wow, I wasn't aware of that option. Sure thing.